### PR TITLE
Independent town creation

### LINKED ISSUE #36: "New villages emerge" option doesn't work
How to reproduce:
* Start a new game with "New villages emerge" option set to "no" and "Independent villages" set to "few".
* You start with 7 independent villages.
* Set speed to high.
* New villages spawn until there are 10 villages on the map. 

Expected behavior:
* New independent villages should not spawn when "New villages emerge"  is set to "no".

When villages are captured new villages keep spawning so that the total amount is about 10 independent villages at all times. This makes it impossible to run out of independent villages.

### PR DESCRIPTION
**Pull request by MicroVirus**
The first patch fixes #36 where independent towns spawn regardless of the configuration option.

The second patch is optional and up for debate. Rather than using a fixed value of max. 10 independent villages, it now uses the number given by the start amount. This will be less for Few (7), and more for More (15) and Many (30), where for many this is a huge increase.
One the one hand, it is more intuitive that the number of villages setting is indicative for the entire game. On the other hand, the game might be better balanced for the amount of 10.

### CODE REVIEW

**Comment by Omokoii** [comment]
Nice, thanks for quick fix. One variant for the second patch could be to make them spawn up to "starting amount" / 2. That way you always have to consume a few villages before the respawn starts. Just a thought since I play with no respawns anyway :)

**Comment by the3dfxdude** [comment]
I've been studying this function in 7kaa and 7k2. This fix appears in 7k2, but..... I think the fix is wrong. The purpose of the function was to help clean up rebels or other "wanderers", and if new_independent_town_emerge is disabled, and the condition is added then this disables the cleanup mechanism.

I understand that this was a game "feature", but really, does it benefit anyone to turn this off? There are already heavy limits placed on emerging "if( allTotalPop > 1000 )" so it negates pretty much any changes to controls such as "if( independentTownCount >= 10 )". I do see strong incentives to keep the emerging on for the early game pop growth, and keep the limits in to prevent a late game explosion.

I want to study more to see if the wanderers are cleaned up any other way, to deal with the late game.

**Comment by J-Moravec** [comment]
So we need to solve two problems:
1. auto creating new villages even when setting is disabled
2. rebel cleanup. -- should be able to create new village even when n.v.e. is disabled.

It seems to me that in 7K2, rebels have serious problem. They wander aimlessly, they often try to go to captured villages of same nationality, they can't enter of course, so after a while, they pick another target. Soon after that, they start attacking random targets.

Nothing like that happens in 7K.

On top of that, I think that when we have a button "disable new village emergence" (I hope that I wrote that right), it should do exactly that. It is OK in 7K, as breeding rate is high enough and capturing villages and moving those people to your center is long and difficult process anyway.

In 7K2, due to low breeding rate, you basically have to capture new villages because it is the sole source of people (for economy, as you should mostly build war machines and hire troops from tavern, using your own people for war is sort of last ditch, which is sad). The ability to create wagon and thus cheaply (it doesn't cost that much of loyalty and villages in 7K2 tends to rebel less often, or so it seems) move population is another sort of... problem. So yeah, my oppinion would be to repair button so it does what it has written on it and rather rebalance the problem with low population growth (and suicidal AI and AI that surrender to other AIs).

**Comment by J-Moravec** [comment]
Are wanderers used anywere else? (as: actually represented in game) or they are only virtual, basically some counter?

btw. I saw some behaviour, new people emerging from map borders and so, even wagon trains, but in 7K, new villages just... appeared. Or were those dozens of neutral people wandering from village to village in 7K actual people and not spies?

gosh, this is more discussion for forum...

**Comment by the3dfxdude** [comment]
When the total pop reaches 1000, or 10 independent villages exist, then what you are saying happens in 7K as well. Maybe it isn't as bad as in 7K2, because yes people can appear at the borders.

I have nothing against having such an option, and will allow a fix in. But as of right now, I would recommend against turning it off in a regular game without looking at the independent AI some more.

As far as the emerging town function, I looked at it further, and the wandering units is a virtual random sampling of each race. They aren't actually playable units. But I wouldn't discount the emerging towns as a sort of vacuum cleaner that gets turned off late game.

**Comment by J-Moravec** [comment]
Ok, I don't understand you anymore. Should we move it to forum or continue here?

Eg. what do you mean by vacuum cleaner? If in 7k, the only neutral moving units are spies and rebels have their own settling function that ignores independent-town-emergence, what vacuum cleaner you have on mind? (in 7K2, the situation could be different)

**Comment by MicroVirus** [comment]
Note: the following is true for 7kaa. I do not know if it also holds for 7k2.

> The purpose of the function [`think_new_independent_town`] was to help clean up rebels or other "wanderers" [..]

I don't understand what you mean by this. The purpose of this function is to ensure an influx of population (and by extension, allowing new kingdoms to emerge, which require an independent town).
I don't really understand how you mean clean-up; I assume you refer to the `race_wander_pop_array`? This array is 'created' specifically for this function. This function is primarily responsible for increasing the virtual 'wandering population' at random, but only when the prerequisites (less than 10 towns, total pop. less than 1000) are met. Then, as a 'fun' addition, the only other way to add wandering population is by resigning units: if you remove a unit by clicking the 'X', or a units loyalty drops to 0 and it disappears, it adds *two* wanderers of the same race. The code explicitly mentions that this is a feature:

    @@ Unit::resign(int remoteAction):
    
    //---- increase the wandering count when a unit is disbanded ----//
    if( race_id )
        town_array.race_wander_pop_array[race_id-1] += 2;     // disbanding one resulted in two wandering units to make the effect more significant

The AI can also decide to resign some of its units, but does so only when it is completely unable to think of an action for the unit.
**TL;DR:** `think_new_independent_town` increases and depletes the `race_wander_pop_array` in order to create independent towns. As a fun addition, resigning units also adds to the wandering population.

So, at least in 7kaa, there is no occurrence of rebels in this story.

**Comment by the3dfxdude** [comment]
While I am seeing some differences (trade-offs in effect) in 7KAA vs 7K2, I'll just leave that as a study for another time.

My basic thought is that if the world has not reached its pop and town cap, that generally an independent unit that is orphaned (whether by betraying or rebelling, or even an independent spy of another player) will automatically try to find a town to settle. Despite that the think_new_independent_town only looks at a virtual sampling, it keeps track of certain races that had resigned units, with some randomness. If the criteria is met for emerging a town, then the game will create a new town with space again. This feeds the early game pop growth, and helps keep open space for receiving the active independent wanderers.

Here are some functions that go along with the clean up, not directly related to think_new_independent_town, but serves the same purpose. Think that if the game is keeping some space open until it reaches the cap, these following functions will generally do something. think_independent_unit() will allow the unit to resign() if a town cannot be found, but the others don't directly do this and may actually mean the unit can get stuck and do nothing.

Rebel::think_new_action()
Unit::process_rebel()
Unit::think_independent_unit()

In the late game, I can see potential for trouble hitting a cap. Turning off emerging towns makes the trouble worse in certain functions, since it will prevent growth altogether before even reaching 1000 world pop, and you will have maxed out town pop @ 60 very quickly. Rebel::think_new_action() does seems to only apply to rebel town creation (completely separate cleanup), but Unit::process_rebel() does not seem to require rebel towns to settle in.

**Comment by MicroVirus** [comment]
The game has the option for turning New Independent Villages on and off. Obviously, when turned off the population growth is much more restricted and will create a different gameplay experience. However, this is exactly what was meant and desired by the user when turning this option off. Therefore, this option should be supported and work properly, both when turned on and off.

I've looked through all the rebel code and none of it will break if there are currently no towns to assign to; there are always alternatives present.
This is as expected (that it doesn't break), because the population cap of 1000 is easily reached and most of us have played countless games without independent towns for a many number of in-game years -- in effect, it has been highly play-tested over the years.

If I understand you correctly, the question you have right now is: "Will disallowing villages to spawn altogether cause extra problems?".
What we know is that late-game, regardless of settings, we always have an absence of independent towns. So with this option on or off things will be the same. Therefore, not allowing this option will not fix anything: any issues that are present will need to be addressed separately.
What I conclude from this is that the independent towns do not function as a clean-up. They are usable as an option, but alternatives exist for rebels and independent units.

I have not seen a single issue with the code where it assumes that an independent village, or even the space for such, exists. The code always assumes the worst can happen, in this regard.

**Comment by MicroVirus** [comment]
The game has the option for turning New Independent Villages on and off. Obviously, when turned off the population growth is much more restricted and will create a different gameplay experience. However, this is exactly what was meant and desired by the user when turning this option off. It's nice to give players ways to customise their games as they see fit. Therefore, this option should be supported and work properly, both when turned on and off.

I've looked through all the rebel code in OREBEL.cpp and none of it will break if there are currently no towns to assign to; there are always alternatives present.
To be honest, I haven't seen any issues with the absence of towns; the population cap of 1000 is easily reached and most of us have played countless games without independent towns for a many number of in-game years. You could say that it has been highly play-tested over the years.

If I understand you correctly, the question you have right now is: "Will disallowing villages to spawn altogether cause extra problems?".
What we know is that late-game, regardless of settings, we always have an absence of independent towns. So with this option on or off things will be the same. Therefore, if there are any issues present then we will need to address them, regardless of this option.

As far as I can see the independent towns do not function as a clean-up. They are usable as an option, but alternatives always exist for rebels and independent units. I could've missed something, of course, but I have not seen an issue with the code where it assumes that an independent village, or even the space for such, exists. The code always assumes the worst can happen, in this regard. Between this and that I have not observed adverse effects, I think we're in the green here.

I suggest using this option and if issues do arise we handle them as they pop up, thereby improving the game both with this option turned on and off.

EDIT: I rewrote the piece a bit to be truer to the way I meant it. I feel it might have come over as too abrasive, which I did not mean.

**Comment by the3dfxdude** [comment]
Oh I see no problem with the fix going in. I just find it strange that option was supposed to be implemented, never did, and never really needed. I'll try to test and check in the first patch.

**Comment by MicroVirus** [comment]
The funny thing is I always assumed it'd work, but I don't like to play
like that, so I never did try it.

One thing that might happen is that there are scenarios which have this
option off, and with the fix the difficulty of these scenarios changes. I
don't know if this happens.

**Comment by the3dfxdude** [comment]
I looked at this by logging the game actions in the function. Watching the behavior, I think that this is easily described what it does.

If new independent villages are turned on, it keeps 10 independent villages on the map at all times. When it hits 1000 pop (which takes a long time to get to), then the villages are allowed to deplete to zero.

If 'low' is selected for the initial number of independent villages, then 3 villages are created rather quickly :)  Any higher, it will simply deplete to 10.

With the switch turned off, which I kind of like, the villages simply deplete (unless rebels create them). I think this works nicely for multiplayer matches. For the AI, I think that it will have poorer growth, since it will have a limited supply to take from (independent villages is really the only way it knows how), and possibly be more moody for always being behind in pop (like usual) :)

Don't see any reason to raise the limit of villages above 10, since they get created so easy. Maybe a extended config option, but I don't see a reason to change the default.
